toolchains/toolchain_info.bzl

```diff
@@ -0,0 +1,168 @@
+# Copyright 2024 The Bazel Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""All providers for rule-based bazel toolchain config."""
+
+visibility(["public"])
```

Comment:
Do we need to make all of these public on day 1? I would feel more comfortable with only the rules being public for now.

Reply:
We need, at the very least, some subset of these to be exposed on day 1, otherwise no rulesets can use the toolchain.

```
def _compile(ctx):
    toolchain = ctx.toolchains[_LANG_TOOLCHAIN_TYPE].config
    run_action(
        ...,
        action_type = ctx.attr._compile_action[ActionTypeInfo],
    )
```